src/storage/src/render/upsert.rs

```diff
@@ -295,10 +294,23 @@ where
         key_indices_sorted.clone(),
         &upsert_envelope.key_indices,
     );
+
+    // It's very important to hash the right thing. We have nested `Option` and
+    // `Result` here. And, for example, `Some(key).hashed()` is not the same as
+    // `key.hashed()`.
+    //
+    // We make sure  to hash the same thing for both pervious updates and new
+    // updates.
+    //
+    // Also: this problem was only showing up when trying to use upsert-style
+    // sources with multiple storaged workers.
     let result_stream = stream.binary_frontier(
         &previous_ok.inner,
-        Exchange::new(move |DecodeResult { key, .. }| key.hashed()),
-        Exchange::new(|((key, _v), _t, _r)| Ok::<_, Infallible>(key).hashed()),
+        Exchange::new(move |DecodeResult { key, .. }| match key {
+            None => 0,
```

Comment:
note: if we wanted to be fancy we could use sth like `source_id.hashed()` instead, but that would make this change bigger because I have to plumb that through.